File: leetcode/go/1404-numberOfStepsToReduceANumberInBinaryRepresentationToOne.go
//go:build ignore

package main

import "fmt"

func numSteps(s string) int {
	res, reminder := 0, 0

	for i := len(s) - 1; i > 0; i-- {
		bit := int(s[i] - '0') + reminder

		if bit == 1 {
			res += 2
			reminder = 1
		} else {
			res += 1
		}
	}

	return res + reminder
}

func main() {
	s := "1101"
	fmt.Print(numSteps(s))
}

File: leetcode/python/1404-numberOfStepsToReduceANumberInBinaryRepresentationToOne.py
class Solution:
    def bit2dec(self, s: str):
        dec = 0
        for bit in s:
            dec <<= 1
            if bit == "1":
                dec += 1
        return dec
    
    def numSteps(self, s: str) -> int:
        dec = self.bit2dec(s)
        res = 0
        
        while dec != 1:
            res += 1
            dec = (dec + 1) if dec % 2 == 1 else (dec >> 1)
            
        return res        


def main():
    s = Solution()
    bits = "1101"
    print(s.numSteps(bits))
    
if __name__ == "__main__":
    main()
    